src/relax/transform/to_mixed_precision.cc

```diff
@@ -0,0 +1,535 @@
+        // arg is Constant
+        const auto* sinfo = GetStructInfoAs<TensorStructInfoNode>(args[i]);
+        ICHECK(sinfo != nullptr) << "Invalid sinfo for constant: " << args[i];
+        ICHECK(to[i].IsLeaf()) << "Invalid dtype for constant: " << args[i];
+        Var casted = builder_->Emit(astype(GetRef<Constant>(constant), to[i].LeafValue()->dtype));
+        new_args->push_back(casted);
+      } else if (const auto* tuple = args[i].as<TupleNode>()) {
+        // arg is Tuple
+        ICHECK(to[i].IsNested()) << "Invalid dtype for tuple: " << args[i];
+        std::vector<Expr> new_fields;
+        CastArgsTo(tuple->fields, to[i].NestedArray(), &new_fields);
+        new_args->push_back(Tuple(new_fields));
+      } else {
+        LOG(FATAL) << "Unsupported argument type: " << args[i]->GetTypeKey();
+      }
+    }
+  }
+
+  void CastArgsTo(Array<Expr> args, DataType to, std::vector<Expr>* new_args) {
+    std::vector<NTensorStruct> to_arr;
+    for (const Expr& arg : args) {
+      to_arr.push_back(NTensorStructFrom(arg, to));
+    }
+    CastArgsTo(args, to_arr, new_args);
+  }
+
+  bool AllFp16Available(const Array<Expr>& args) {
+    for (const Expr& arg : args) {
+      // arg is either a var or a constant or a tuple
+      if (const auto* var_node = arg.as<VarNode>()) {
+        // arg is a var
+        auto it = realized_var_.find(GetRef<Var>(var_node));
+        ICHECK(it != realized_var_.end()) << "Var not found: " << arg;
+        Var realized = it->second;
+        NTensorStruct realized_sinfo = NTensorStructFrom(realized);
+        NTensorStruct realized_fp16 = NTensorStructFrom(realized, fp16_);
+        if (!NTensorStructEqual()(realized_fp16, realized_sinfo)) {
+          return false;
+        }
+      } else if (const auto* tuple_node = arg.as<TupleNode>()) {
+        if (!AllFp16Available(tuple_node->fields)) {
+          return false;
+        }
+      }
+      // constant can always be casted, so we don't need to check it
+    }
+    return true;
+  }
+
+  void VisitBinding_(const VarBindingNode* binding, const CallNode* call_node) override {
+    Optional<Array<ObjectRef>> dtype_convert_info = GetDTypeConvertInfo(call_node, output_dtype_);
+    if (!dtype_convert_info.defined()) {
+      const auto* op = call_node->op.as<OpNode>();
+      if (op == nullptr || op->name != "relax.wrap_param") {
+        ExprMutator::VisitBinding_(binding, call_node);
+        return;
+      }
+      // wrap_param
+      const auto* constant = call_node->args[0].as<ConstantNode>();
+      ICHECK(constant != nullptr) << "Invalid wrap_param: " << GetRef<Call>(call_node);
+      ReEmitBinding(binding, builder_->Normalize(GetRef<Constant>(constant)));
+      return;
+    }
+    // var: NTensorStruct = Call(Op, [args], attr)
+    Array<ObjectRef> res = dtype_convert_info.value();
+    MixedTypePolicy policy = static_cast<MixedTypePolicy>(Downcast<Integer>(res[0])->value);
```

Comment:
Suppose one operator might have multiple outputs with different types? (e.g. `max` returns a value tensor with float type and an indices tensor with int type). How to describe the behavior?

Reply:
Generally, we can extend the signature to return NTensorType for each of its output.

Layout conversion already follows this behavior https://github.com/mlc-ai/relax/blob/relax/src/relax/transform/infer_layout_utils.h#L49

Because for ops like BN, it will return 3 tensors, each with a different layout conversion rule.

But in AMP currently, we don't have such use cases.